Snowflake terraform upgrade to v1.0.1

ram-kishore-odi opened the pull request:
Hi @ian-r-rose, Can you please share your feedback after reviewing the information below  ?

Attached are the plans before and after the import of 5 service users. 

Surprisingly the plan still shows 5 resources need to be replaced. [1.0-plan-dev-after-import.txt](https://github.com/user-attachments/files/18294763/1.0-plan-dev-after-import.txt)

The inital plan showed that the resources will be destroyed.
[1.0-plan-dev-before-removal.txt](https://github.com/user-attachments/files/18294762/1.0-plan-dev-before-removal.txt)

I did not run into any issues while removing or importing the resorces
Here are the sample commands I ran - 
terraform state rm module.elt.snowflake_user.fivetran
terraform import module.elt.snowflake_service_user.fivetran FIVETRAN_SVC_USER_DEV

ian-r-rose reviewed (approved):
These changes look good to me @ram-kishore-odi. If you look at the plan you posted, you can see which attribute forces the removal (in this case, it looks like `user_type`). I suspect this can be fixed without destroying the user by altering them in Snowflake.

ram-kishore-odi commented:
Thank you @ian-r-rose ! I will use the alter command change user to TYPE = SERVICE and LEGACY_SERVICE for Sentinel.
I checked Snowflake and was surprised to see that the User Type is not populated both for human and service users in the portal. So will check for one service user and then proceed to update others

ian-r-rose commented:
> I checked Snowflake and was surprised to see that the User Type is not populated both for human and service users in the portal. 

I think this is because it's a new parameter, and did not exist when these users were created

ram-kishore-odi commented:
Looks like your suggestion is working. Thank you !

I updated for FiveTran service user and now it says. So I may still need to apply the changes after running manual alter commands in SF.

module.elt.snowflake_service_user.fivetran will be **updated in-place**
  ~ resource "snowflake_service_user" "fivetran" {
      ~ disabled                                      = "false" -> "default"
        id                                            = "FIVETRAN_SVC_USER_DEV"
      - login_name                                    = (sensitive value) -> null
      + mins_to_unlock                                = -1
        name                                          = "FIVETRAN_SVC_USER_DEV"
      ~ show_output                                   = [
          - {
              - comment                 = "Service user for Fivetran"
              - created_on              = "2023-05-31 09:23:34.136 -0700 PDT"
              - default_role            = "LOADER_DEV"
              - default_warehouse       = "LOADING_XS_DEV"
              - disabled                = false
              - expires_at_time         = "0001-01-01 00:00:00 +0000 UTC"
              - ext_authn_duo           = false
              - has_mfa                 = false
              - has_password            = false
              - has_rsa_public_key      = false
              - last_success_login      = "0001-01-01 00:00:00 +0000 UTC"
              - locked_until_time       = "0001-01-01 00:00:00 +0000 UTC"
              - login_name              = "FIVETRAN_SVC_USER_DEV"
              - must_change_password    = false
              - name                    = "FIVETRAN_SVC_USER_DEV"
              - owner                   = "USERADMIN"
              - snowflake_lock          = false
              - type                    = "SERVICE"
                # (10 unchanged attributes hidden)
            },
        ] -> (known after apply)
        # (65 unchanged attributes hidden)
    }

ian-r-rose commented:
Great, looks like smooth sailing!

ram-kishore-odi commented:
Hello @ian-r-rose - Please see the final updates below - 

- Successfully applied changes to the development environment. 
- Added both Mac and Linux provider binaries in the terraform lock files (Dev & Prd)
- Updated production configuration
- Will apply changes to PRD from main branch (after merging right away)
- Production plan shows the same changes as that of Dev.

******** Dev Apply -- Sucessful *********
Plan: 0 to add, 5 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.elt.snowflake_legacy_service_user.sentinel: Modifying... [id=SENTINEL_SVC_USER_DEV]
module.elt.snowflake_legacy_service_user.sentinel: Modifications complete after 1s [id=SENTINEL_SVC_USER_DEV]
module.elt.snowflake_service_user.github_ci: Modifying... [id=GITHUB_ACTIONS_SVC_USER_DEV]
module.elt.snowflake_service_user.github_ci: Modifications complete after 1s [id=GITHUB_ACTIONS_SVC_USER_DEV]
module.elt.snowflake_service_user.dbt: Modifying... [id=DBT_CLOUD_SVC_USER_DEV]
module.elt.snowflake_service_user.dbt: Modifications complete after 0s [id=DBT_CLOUD_SVC_USER_DEV]
module.elt.snowflake_service_user.airflow: Modifying... [id=MWAA_SVC_USER_DEV]
module.elt.snowflake_service_user.airflow: Modifications complete after 1s [id=MWAA_SVC_USER_DEV]
module.elt.snowflake_service_user.fivetran: Modifying... [id=FIVETRAN_SVC_USER_DEV]
module.elt.snowflake_service_user.fivetran: Modifications complete after 0s [id=FIVETRAN_SVC_USER_DEV]

Apply complete! Resources: 0 added, 5 changed, 0 destroyed.
